src/storage/src/render/upsert.rs

```diff
@@ -200,19 +215,87 @@ fn evaluate(
     Ok(Some(row_buf.clone()))
 }
 
+/// Given a stream of rows and a description of the columns that form their key,
+/// produce a stream of keys and thinned values.
+fn extract_kv<G: Scope>(
+    records: Collection<G, Row, Diff>,
+    key_indices_sorted: Vec<usize>,
+    key_indices: &[usize],
+) -> Collection<G, (Row, Row), Diff> {
+    debug_assert!({
+        let mut verified_sorted = key_indices.to_vec();
+        verified_sorted.sort_unstable();
+        key_indices_sorted == verified_sorted
+    });
+    let key_cols_are_sorted = &key_indices_sorted == key_indices;
+
+    let mut row_buf = Row::default();
+    // If the key columns are in order, we can pack them directly
+    // into `key_row_buf` while scanning the original row.
+    //
+    // Otherwise, we need to put them into `key_dv` and then invert the permutation
+    // before packing.
+    let mut key_row_buf = Row::default();
+    let mut key_dv = DatumVec::default();
+    let key_unsort_permutation = inverse_argsort(key_indices);
+    records.map(move |row| {
+        let mut row_packer = row_buf.packer();
+        let mut key_row_packer = key_row_buf.packer();
+        let values = &mut row.iter();
+        let mut next_idx = 0;
+        let mut key_dv = key_dv.borrow();
+        for &key_idx in key_indices_sorted.iter() {
+            // First, push the datums that are before `key_idx`
+            row_packer.extend(values.take(key_idx - next_idx));
+            // Then, add the key field to whichever buffer we're using for the key
+            let key_datum = values.next().unwrap();
+            if key_cols_are_sorted {
+                key_row_packer.push(key_datum);
+            } else {
+                key_dv.push(key_datum);
+            }
+            next_idx = key_idx + 1;
+        }
+        // Finally, push any columns after the last key index
+        row_packer.extend(values);
```

Comment:
is this just like `thin`, but also produces the `key` `Row` as well?

Reply:
Yep!